gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/KillDagProc.java

```diff
@@ -0,0 +1,113 @@
+  private final KillDagTask killDagTask;
+
+  public KillDagProc(KillDagTask killDagTask) {
+    this.killDagTask = killDagTask;
+  }
+
+  protected CancelEntity initialize(DagManagementStateStore dagManagementStateStore) throws IOException {
+    Dag<JobExecutionPlan> dagToCancel = dagManagementStateStore.getDag(this.killDagTask.getDagId().toString());
+    List<Dag.DagNode<JobExecutionPlan>> dagNodesToCancel = new ArrayList<>(dagManagementStateStore.getJobs(this.killDagTask.getDagId().toString()));
+    return new CancelEntity(dagToCancel, dagNodesToCancel);
+  }
+
+  @Override
+  public Optional<Dag<JobExecutionPlan>> act(DagManagementStateStore dagManagementStateStore, CancelEntity cancelEntity) throws IOException {
+    if (cancelEntity.dagToCancel == null || cancelEntity.dagNodesToCancel.isEmpty()) {
+      log.warn("No dag with id " + this.killDagTask.getDagId() + " found to kill");
+      return Optional.empty();
+    }
+    for (Dag.DagNode<JobExecutionPlan> dagNodeToCancel : cancelEntity.dagNodesToCancel) {
+      cancelDagNode(cancelEntity.dagToCancel, dagNodeToCancel, dagManagementStateStore);
+    }
+    return Optional.of(cancelEntity.dagToCancel);
+  }
+
+  private void cancelDagNode(Dag<JobExecutionPlan> dagToCancel, Dag.DagNode<JobExecutionPlan> dagNodeToCancel, DagManagementStateStore dagManagementStateStore) throws IOException {
+    Properties props = new Properties();
+    try {
+      if (dagNodeToCancel.getValue().getJobFuture().isPresent()) {
+        Future future = dagNodeToCancel.getValue().getJobFuture().get();
+        String serializedFuture = DagManagerUtils.getSpecProducer(dagNodeToCancel).serializeAddSpecResponse(future);
+        props.put(ConfigurationKeys.SPEC_PRODUCER_SERIALIZED_FUTURE, serializedFuture);
+        sendCancellationEvent(dagNodeToCancel.getValue());
+      }
+      if (dagNodeToCancel.getValue().getJobSpec().getConfig().hasPath(ConfigurationKeys.FLOW_EXECUTION_ID_KEY)) {
+        props.setProperty(ConfigurationKeys.FLOW_EXECUTION_ID_KEY,
+            dagNodeToCancel.getValue().getJobSpec().getConfig().getString(ConfigurationKeys.FLOW_EXECUTION_ID_KEY));
+      }
+      DagManagerUtils.getSpecProducer(dagNodeToCancel).cancelJob(dagNodeToCancel.getValue().getJobSpec().getUri(), props);
+      dagToCancel.setFlowEvent(TimingEvent.FlowTimings.FLOW_CANCELLED);
+      dagToCancel.setMessage("Flow killed by request");
+      dagManagementStateStore.removeDagActionFromStore(this.killDagTask.getDagAction());
+    } catch (ExecutionException | InterruptedException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private void sendCancellationEvent(JobExecutionPlan jobExecutionPlan) {
+    Map<String, String> jobMetadata = TimingEventUtils.getJobMetadata(Maps.newHashMap(), jobExecutionPlan);
+    this.eventSubmitter.getTimingEvent(TimingEvent.LauncherTimings.JOB_CANCEL).stop(jobMetadata);
+    jobExecutionPlan.setExecutionStatus(CANCELLED);
+  }
+
+  @AllArgsConstructor
```

Comment:
could all the members be `final`?

why not encapsulate behavior here such as:
```
    for (Dag.DagNode<JobExecutionPlan> dagNodeToCancel : cancelEntity.dagNodesToCancel) {
      cancelDagNode(cancelEntity.dagToCancel, dagNodeToCancel, dagManagementStateStore);
    }
```
as
```
/** @returns how many canceled; @throws ...??? */
public int cancelAllDagNodes(DMStateStore stateStore);
```

Reply:
let me revisit this in next PRs